// babel-plugin-angular2-annotations/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (_ref) {
  var t = _ref.types;

  var helper = (0, _babelRttsHelper2.default)({ types: t }, 'assert');

  return {
    visitor: {
      ClassDeclaration: function ClassDeclaration(path, pass) {
        var node = path.node;
        var classRef = node.id;
        var classBody = node.body.body;

        // Create additional statements for parameter decorators and types.
        var decorators = [];
        var types = [];
        classBody.forEach(function (bodyNode) {
          if (bodyNode.type === 'ClassMethod' && bodyNode.kind === 'constructor') {
            decorators = parameterDecorators(bodyNode.params, classRef);
            types = parameterTypes(bodyNode.params, classRef);
          } else if (bodyNode.type === 'ClassProperty' && bodyNode.value === null && !bodyNode.static) {
            // Handle class property without initializer.
            // https://github.com/jeffmo/es-class-fields-and-static-properties
            bodyNode.value = t.memberExpression(t.thisExpression(), bodyNode.key);
          }
        });
        var additionalStatements = [].concat(_toConsumableArray(decorators), _toConsumableArray(types)).filter(Boolean);

        // If not found, do nothing.
        if (additionalStatements.length === 0) {
          return undefined;
        }

        // Append additional statements to program.
        var parent = path.parent;
        if (parent.type === 'ExportNamedDeclaration' || parent.type === 'ExportDefaultDeclaration') {
          // The class declaration is wrapped by an export declaration.
          path.parentPath.insertAfter(additionalStatements);
        } else {
          path.insertAfter(additionalStatements);
        }
      }
    }
  };

  // Returns an array of parameter decorator call statements for a class.
  function parameterDecorators(params, classRef) {
    var decoratorLists = params.map(function (param, i) {
      var decorators = param.decorators;
      if (!decorators) {
        return [];
      }
      // Delete parameter decorators because they are invalid in vanilla babel.
      // They might be just ignored in generator though.
      param.decorators = null;

      return decorators.map(function (decorator) {
        var call = decorator.expression;
        var args = [classRef, t.identifier('null'), t.identifier(i.toString())];
        return t.expressionStatement(t.callExpression(call, args));
      });
    });
    // Flatten.
    return Array.prototype.concat.apply([], decoratorLists);
  }

  // Returns an array of define 'parameters' metadata statements for a class.
  // The array may contain zero or one statements.
  function parameterTypes(params, classRef) {
    var types = params.map(function (param) {
      var annotation = param.typeAnnotation && param.typeAnnotation.typeAnnotation;
      if (!annotation) {
        return null;
      }
      return helper.typeForAnnotation(annotation);
    });
    if (!types.some(Boolean)) {
      return [];
    }
    return [defineMetadata('design:paramtypes', t.arrayExpression(types), classRef)];
  }

  // Returns an AST for define metadata statement.
  function defineMetadata(key, value, target) {
    return t.expressionStatement(t.callExpression(t.memberExpression(t.identifier('Reflect'), t.identifier('defineMetadata')), [t.stringLiteral(key), value, target]));
  }
};

var _babelRttsHelper = require('babel-rtts-helper');

var _babelRttsHelper2 = _interopRequireDefault(_babelRttsHelper);

var _patch = require('./patch');

var _patch2 = _interopRequireDefault(_patch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

(0, _patch2.default)();
